src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,456 @@
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editor.Implementation.AutomaticCompletion;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Commanding;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor.Commanding.Commands;
+using Microsoft.VisualStudio.Text.Operations;
+using Microsoft.VisualStudio.Utilities;
+using VSCommanding = Microsoft.VisualStudio.Commanding;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.CompleteStatement
+{
+    /// <summary>
+    /// When user types <c>;</c> in a statement, closing delimiters and semi-colon are added and caret is placed after the semicolon
+    /// </summary>
+    [Export(typeof(VSCommanding.ICommandHandler))]
+    [ContentType(ContentTypeNames.CSharpContentType)]
+    [Name(nameof(CompleteStatementCommandHandler))]
+    internal sealed class CompleteStatementCommandHandler : IChainedCommandHandler<TypeCharCommandArgs>
+    {
+        private readonly ITextUndoHistoryRegistry _undoHistoryRegistry;
+        private readonly IEditorOperationsFactoryService _editorOperationsFactoryService;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CompleteStatementCommandHandler(ITextUndoHistoryRegistry undoHistoryRegistry, IEditorOperationsFactoryService editorOperationsFactoryService)
+        {
+            _undoHistoryRegistry = undoHistoryRegistry;
+            _editorOperationsFactoryService = editorOperationsFactoryService;
+        }
+
+        public string DisplayName => CSharpEditorResources.Complete_statement;
+
+        public void ExecuteCommand(TypeCharCommandArgs args, Action nextCommandHandler, CommandExecutionContext executionContext)
+        {
+            if (args.TypedChar != ';')
+            {
+                nextCommandHandler();
+                return;
+            }
+
+            var caret = args.TextView.GetCaretPoint(args.SubjectBuffer);
+            if (!caret.HasValue)
+            {
+                nextCommandHandler();
+                return;
+            }
+
+            var document = caret.Value.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                nextCommandHandler();
+                return;
+            }
+
+            var root = document.GetSyntaxRootSynchronously(executionContext.OperationContext.UserCancellationToken);
+            var caretPosition = caret.Value.Position;
+
+            var token = GetToken(root, caretPosition, caret.Value);
+
+            var currentNode = token.Parent;
+            // if cursor is right before an opening delimiter, make sure you start with node outside of delimiters
```

Comment:
can this be clarified at all?

Reply:
👍 I'll update the comment. The case it covers is this:

```csharp
obj.ToString$$()
```

For this case, the token we find is the `(`, but we aren't actually inside the argument list so we don't want to analyze from that position.